system/jlib/jdebug.cpp

```diff
@@ -1706,61 +1754,360 @@ class CProcessMonitor
-        __uint64 rxerrors;
-        __uint64 rxdrops;
-        __uint64 txbytes;
-        __uint64 txpackets;
-        __uint64 txerrors;
-        __uint64 txdrops;
-    };
+        unsigned reads = 0;
+        if (sscanf(ln, "%4d %4d %31s %u", &pi.major, &pi.minor, pi.name, &reads) == 4)
+        {
+            unsigned p = 0;
+            while ((p<nparts) && (partition[p].major != pi.major || partition[p].minor != pi.minor))
+                p++;
+            if ((p==nparts) && reads && isDisk(pi.major,pi.minor))
+            {
+                nparts++;
+                partition = (part_info *)realloc(partition,nparts*sizeof(part_info));
+                partition[p] = pi;
+            }
+        }
+    }
+    fclose(diskfp);
+}
+
+
+
+void OsDiskInfo::initMajorMinor()
+{
+#ifdef __linux__
+    // MCK - wish libblkid could do this ...
+    // Another way might also be to look for:
+    //   /sys/block/sd*
+    //   /sys/block/nvme*
+    // and match those with entries in /proc/diskstats
+    StringBuffer cmd("lsblk -o TYPE,MAJ:MIN --pairs");
+    Owned<IPipeProcess> pipe = createPipeProcess();
+    if (pipe->run("list disks", cmd, nullptr, false, true, true, 8192))
+    {
+        StringBuffer output;
+        Owned<ISimpleReadStream> pipeReader = pipe->getOutputStream();
+        readSimpleStream(output, *pipeReader);
+        unsigned exitcode = pipe->wait();
+        if ( (exitcode == 0) && (output.length() > 0) )
+        {
+            StringArray lines;
+            lines.appendList(output, "\n");
+            ForEachItemIn(idx, lines)
+            {
+                // line: TYPE="disk" MAJ:MIN="259:0"
+                unsigned majnum, minnum;
+                if (2 == sscanf(lines.item(idx), "TYPE=\"disk\" MAJ:MIN=\"%u:%u\"", &majnum, &minnum))
+                {
+                    unsigned mm = (majnum<<16)+minnum;
+                    diskMajorMinor.appendUniq(mm);
+                }
+            }
+        }
+        else
+        {
+            StringBuffer outputErr;
+            Owned<ISimpleReadStream> pipeReaderErr = pipe->getErrorStream();
+            readSimpleStream(outputErr, *pipeReaderErr);
+            if (outputErr.length() > 0)
+                WARNLOG("WARNING: Pipe: output: %s", outputErr.str());
+        }
+    }
+#endif // __linux__
+}
 
-    struct part_info 
+unsigned OsDiskInfo::mapPartition(unsigned major, unsigned minor) const
+{
+    for (unsigned p = 0; p < nparts; p++)
     {
-        unsigned int major; 
-        unsigned int minor; 
-        char name[32];
-    };
-    
+        if (partition[p].major == major && partition[p].minor == minor)
+            return p;
+    }
+    return (unsigned)-1;
+}
+
+
+static Singleton<OsDiskInfo> globalOsDiskInfo;
+MODULE_INIT(INIT_PRIORITY_JDEBUG2)
+{
+    return true;
+}
+MODULE_EXIT()
+{
+    delete globalOsDiskInfo.queryExisting();
+}
+
+const OsDiskInfo & queryGlobalOsDiskInfo()
+{
+    return *globalOsDiskInfo.query([] { return new OsDiskInfo; });
```

Comment:
I need to understand this line

Reply:
It is a way of implementing code like the following:

  if (!ptr)
  {
     CriticialBlock block(cs);
     if (!ptr)
        ptr = new OsDiskInfo;
  }
  return ptr;

ensuring that atomics are used correctly and not re-evaluated.